### PR TITLE
Wrap ilCtrlStructureReader into an Objective

### PR DESCRIPTION
**Pull request by nhaagen**
Use ilCtrlStructureReader with cli-setup's Agent/Objective.
This makes very little changes to the CtrlStructureReader itseltf and thus does not break common setup-routines.

### CODE REVIEW

File: Services/UICore/classes/Setup/class.ilCtrlStructureSetupAgent.php
Comment (klees) [inline Services/UICore/classes/Setup/class.ilCtrlStructureSetupAgent.php]: Let's call this `UICoreSetupAgent`. This might want to do something more than just control-structure...

File: Services/Database/classes/Setup/class.ilDatabaseSetupAgent.php
Comment (klees) [inline Services/Database/classes/Setup/class.ilDatabaseSetupAgent.php:64]: :+1:

File: Services/UICore/classes/Setup/class.ilCtrlStructureSetupAgent.php
Comment (klees) [inline Services/UICore/classes/Setup/class.ilCtrlStructureSetupAgent.php]:
```suggestion
		return new \ilCtrlStructureStoredObjective($this->ctrl_reader);
```

I see no harm in reloading the control structure on every update. This will get rid of the requirement that devs need to explicitly do that.

File: Services/UICore/classes/Setup/class.ilCtrlStructureStoredObjective.php
Comment (klees) [inline Services/UICore/classes/Setup/class.ilCtrlStructureStoredObjective.php]: I think we should just remove this path (and the according method). I rather see this going into the "artifact builder" direction, so this will also not be required in the future (right?).

File: setup/classes/class.ilCtrlStructureReader.php
Comment (klees) [inline setup/classes/class.ilCtrlStructureReader.php]: Please remove this global as you did with `$ilDB`. Are there any other?

**Review by klees — APPROVED** [review]
Hi @nhaagen,

I have some minor change requests, but in general this LGTM. I think @alex40724 (as implicit maintainer) should have a look here too.

Best regards!

File: Services/UICore/classes/Setup/class.ilCtrlStructureStoredObjective.php
Comment (nhaagen) [reply Services/UICore/classes/Setup/class.ilCtrlStructureStoredObjective.php]: right.

File: setup/classes/class.ilCtrlStructureReader.php
Comment (nhaagen) [reply setup/classes/class.ilCtrlStructureReader.php]: Where am I to pull $lng from? any hints?
btw - lng ist not used but in an exception-message that asks the user to reload the control-structure. Maybe we can reduce to english only in this case and ditch $lng altogether?

This is the line:
383: duplicate_ctrl#:#Error: Duplicate call structure definition found (Class %s) in files:<br /><br />- %s<br />- %s<br /><br />Please remove the file, that does not belong to the official ILIAS distribution.<br />After that invoke 'Tools' -> 'Reload Control Structure' in the ILIAS Setup.###07 09 2006 new variable

Comment (alex40724) [reply setup/classes/class.ilCtrlStructureReader.php]: Background: It is a common fault of administrators to just unzip ILIAS packages into existing ILIAS code directories. Which leads to the existence of old code files "mostly" overwritten by new code files. But if not included in the new packages, old files will persist. I expect this not only being a problem for the structure reader but also for the artifact reader.

Anyway, I agree that administrators should be able to cope with an English error message, so removing the $lng dependency completely is fine for me.

**Review by alex40724 — APPROVED** [review]

**Comment by alex40724** [comment]
Thanks!

**Comment by klees** [comment]
Thx @alex40724!

**Comment by nhaagen** [comment]
Hey @klees,
these commits are a bit clearer; is this what you had in mind?

**Comment by klees** [comment]
@nhaagen: Yes, thx!

**Pull request merged**
